include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh

```diff
@@ -1208,14 +1253,21 @@ class open_addressing_ref_impl {
                                                         value_type const& expected,
                                                         Value const& desired) noexcept
   {
-    if constexpr (sizeof(value_type) <= 8) {
-      return packed_cas(address, expected, desired);
+    if constexpr (is_perfect_hashing<decltype(probing_scheme_)>::value) {
+      // Do it without atomics.
+      *address = desired;
```

Comment:
@wence- is right, this should be atomic store.

Reply:
Yes, @wence- is right.  It is UB, even if `desired` is the same for all threads writing to the same location.

It's kind of like jaywalking in New York.  Technically, it's illegal.  In practice, everyone is doing it.  If you want, you can avoid jaywalking and stand at the corner and tell everyone next to you, "Look, I am not a criminal because I wait here at the corner!"  But no one will hear you say that because all of them are on the other side of the road already.

Let's look at the pros and cons of the options:

* Don't use atomics
  * Pros:
    * Hash table is now way faster.
    * It works.
  * Cons:
    * It uses UB, which works now and lots of people do it but maybe in the future it won't.
* Use atomics
  * Pros:
    * No UB.  The code is guaranteed to keep working into the future.
    * It works.
  * Cons:
    * The hashing is slower than the competitors so people stop using cuco.

I'm not sure how to deal with this issue.  If we want to be very strict and avoid UB, we can.  And we'll be able to tell all the users that there is no UB in cuco.  But maybe there won't be any users because they all went to go use a faster solution made by the competitor.  (I am the one that made that competitor's solution, ironically.)

There is still some benefit.  Even without removing the atomic here, we still have the removal of the modulo operator and the faster equality operation.  Those are kind of the side-show, though.  The main attraction is removing atomics.

I propose:
1. Get the perfect hashing and perfect probing PRs put into cuco, without this UB but with the modulo and equality optimizations, then
2. Test the performance with/without the UB.
  1. If the performance difference **is not** significant, don't do UB.
  2. If the performance difference **is** significant then...?

What if we let the user decide?  It would be easy to provide a template parameter to the perfect probing called `ub_allowed` that is by default `false`.  And if the user is willing to take the risk then let them.  This makes code that by default has no UB but also cuco's users don't have to go switch to the competitor's solution when they want more performance.

---

It'll be a while until we can measure this.  In the meanwhile, can we get an estimate on how big the performance difference is?  My [TPC-H](https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-h_v3.0.0.pdf) (Q1, page 29) test case is doing 300,000,000 inserts into a hash table where the cardinality of the data is only 4.  (This is actually considered a small test case for TPC-H).  At best case, there are 75,000,000 `atomicCAS` operations going into the same memory location.  We could measure that with `atomicCAS` versus simple write and see what happens.

We should check with all `desired` equal and also all `desired` not equal because, for the `row_hasher` that databases like cuDF use, it's going to be all not equal.

Reply:
> This makes code that by default has no UB but also cuco's users don't have to go switch to the competitor's solution when they want more performance.

I am not a cuco core dev, so perhaps my opinion here carries no real weight, but I would find it very hard to justify a mode in which you can use the library where you can provide _no guarantees_ about whether a result is correct. Particularly, just because you can check that the "no atomics" code works for some test case, is zero guarantee that it is producing the correct code for any other execution.

Reply:
FWIW, the cheapest valid thing to do here is, I think, to use a `relaxed` store. This is a strong (in the memory order sense) operation, rather than a weak one, and therefore removes any UB: if there are different values of `desired` being written, a relaxed store says "I don't care which one I see, or in which order", but I have to see one of them. Whereas a standard store doesn't give the guarantee that you have to see one of the writes.

Reply:
Let's first measure with a relaxed store and see how it goes.

> I would find it very hard to justify a mode in which you can use the library where you can provide no guarantees about whether a result is correct

It's possible that you'll lose customers if you don't at least provide the option because the customer already has the faster version that doesn't use atomics.  Why would they switch to slower code?

Reply:
Perhaps I am too idealistic, but to me, a library which produces the wrong answer (sometimes, you can't know!) very fast is about as useful as a chocolate teapot.

Reply:
Relaxed atomic store is the right way to go.

No library wants to tolerate undefined behavior.